espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/EspressoW3CActionAdapter.kt

```diff
@@ -0,0 +1,158 @@
+            // touch up actions need to be grouped together
+            multiTouchState.updateTouchState(ACTION_UP, sourceId, roundedCoords.x.toLong(), roundedCoords.y.toLong(), globalKeyInputState, button)
+        } else {
+            val xList = listOf(roundedCoords.x.toLong())
+            val yList = listOf(roundedCoords.y.toLong())
+            val androidMotionEvent = AndroidMotionEvent.getMotionEvent(sourceId, uiController)
+            androidMotionEvent.pointerEvent(xList, yList,
+                    ACTION_POINTER_UP, button, pointerType, globalKeyInputState, null, 0)
+            androidMotionEvent.pointerEvent(xList, yList,
+                    ACTION_UP, button, pointerType, globalKeyInputState, null, 0)
+        }
+    }
+
+    @Throws(AppiumException::class)
+    override fun pointerMove(sourceId: String?, pointerType: PointerType?,
+                             currentX: Float?, currentY: Float?, x: Float?, y: Float?,
+                             buttons: Set<Int>?, globalKeyInputState: KeyInputState?) {
+        this.logger.info(String.format("Running pointer move at coordinates: %s %s %s", x, y, pointerType))
+        val roundedCoords = toCoordinates(x, y)
+        if (isTouch(pointerType)) {
+            multiTouchState.updateTouchState(ACTION_MOVE, sourceId, roundedCoords.x.toLong(), roundedCoords.y.toLong(), globalKeyInputState, null)
+            multiTouchState.pointerMove(uiController)
+        } else {
+            AndroidMotionEvent.getMotionEvent(sourceId, uiController)
+                    .pointerMove(listOf(roundedCoords.x.toLong()), listOf(roundedCoords.y.toLong()), pointerType, globalKeyInputState, null)
+        }
+    }
+
+    @Throws(AppiumException::class)
+    override fun pointerCancel(sourceId: String, pointerType: PointerType) {
+        if (isTouch(pointerType)) {
+            multiTouchState.pointerCancel(uiController)
+        } else {
+            AndroidMotionEvent.getMotionEvent(sourceId, uiController).pointerCancel()
+        }
+    }
+
+    @Throws(AppiumException::class)
+    override fun sychronousTickActionsComplete() {
+        multiTouchState.perform(uiController)
+        AndroidLogger.logger.info("Pointer event: Tick complete")
+    }
+
+    override fun getKeyCode(keyValue: String?, location: Int): Int {
+        return AndroidKeyEvent.getKeyCode(keyValue, location)
+    }
+
+    @Throws(AppiumException::class)
+    override fun getElementCenterPoint(elementId: String?): Point {
+        val view = Element.getViewById(elementId)
+        val coords = GeneralLocation.CENTER.calculateCoordinates(view)
+        val point = Point()
+        point.x = Math.round(coords[0])
+        point.y = Math.round(coords[1])
+        return point
+    }
+
+    override fun waitForUiThread() {
+        uiController.loopMainThreadUntilIdle()
+    }
+
+    override fun sleep(duration: Float) {
+        val roundedDuration = Math.round(duration).toFloat()
+        if (duration != roundedDuration) {
```

Comment:
this is always going to be true because it does not make sense to compare two float value for equality